functions/grapher/[slug].ts

```diff
@@ -41,7 +44,49 @@ export const onRequestGet: PagesFunction = async (context) => {
             { ...env, url },
             request.headers.get("if-none-match")
         )
-        .catch((e) => error(500, e))
+        .catch(async (e) => {
+            // Here we do a unified after the fact handling of 404s to check
+            // if we have a redirect in the _grapherRedirects.json file.
+            // This is done as a catch handler that checks for 404 pages
+            // so that the common, happy path does not have to fetch the redirects file.
+            if (e instanceof StatusError && e.status === 404) {
+                console.log("Handling 404 for ", url.pathname)
+                const fullslug = url.pathname.split("/").pop() as string
+
+                const allExtensions = Object.values(extensions)
+                    .map((ext) => ext.replace(".", "\\.")) // for the regex make sure we match only a single dot, not any character
+                    .join("|")
+                const regexForKnownExtensions = new RegExp(
+                    `^(?<slug>.*?)(?<extension>${allExtensions})?$`
+                )
```

Comment:
I think this code will be simpler by assuming that a slug can't contain a dot, i.e. something like `/^(?<slug>[^\s\.]+)(?<extension>\.\S+)?$/`.

In theory we could then check after the fact that the extension is valid (and throw a fitting error), but it's also fine if we're not doing that.

Reply:
Yeah true but I think I like the idea of using a list of known extensions as a communication device that is added to further up in the stack. I'm inclined to leave it as is for now.